File: mssp/examples/019-applicability-is-part-of-the-answer/src/SCL/policy.py
"""Deployment policy: which policy runs, and what an unmeasurable unit means."""
import json
import pathlib

POLICY = json.loads((pathlib.Path(__file__).parent / "policy.json").read_text(encoding="utf-8"))


def policy_name():
    return POLICY["policy"]


def assumption():
    return POLICY["assumes_declarations_are"]


def unmeasurable_is_fatal():
    return POLICY["an_unmeasurable_unit_is"] == "fatal"


def describe():
    return (f'{POLICY["deployment"]}: apply {POLICY["policy"]}, assuming declarations are '
            f'{POLICY["assumes_declarations_are"]}')


def contradicts(measured):
    """Only a MEASURED positive incentive contradicts the assumption.

    A reading that was never taken cannot contradict anything, and must not be
    allowed to look like agreement either.
    """
    if POLICY["assumes_declarations_are"] != "self-penalising":
        return None
    if not measured["applicable"]:
        return None
    if measured["value"] > 0:
        return (f'declaring left {measured["source"]} better off by {measured["value"]} '
                f'record(s) ({measured["suppressed"]} suppressed -> {measured["declared"]} declared)')
    return None

File: mssp/examples/018-who-the-declaration-serves/src/SCL/policy.mjs
import fs from "node:fs";
import path from "node:path";
import { fileURLToPath } from "node:url";

const here = path.dirname(fileURLToPath(import.meta.url));
export const POLICY = JSON.parse(fs.readFileSync(path.join(here, "policy.json"), "utf8"));

export const policyName = () => POLICY.policy;
export const assumption = () => POLICY.assumes_declarations_are;
export const isFatal = () => POLICY.on_contradiction === "fatal";
export const describe = () =>
  `${POLICY.deployment}: apply ${POLICY.policy}, assuming declarations are ${POLICY.assumes_declarations_are}`;

// The assumption, made checkable. A self-penalising declaration must not leave
// the unit that made it better off than suppressing it would have.
export function contradicts(measured) {
  if (POLICY.assumes_declarations_are !== "self-penalising") return null;
  return measured.delta > 0
    ? `declaring left ${measured.source} better off by ${measured.delta} record(s) `
      + `(${measured.suppressed} suppressed -> ${measured.declared} declared)`
    : null;
}
